### PR TITLE
drivers: uart: nrfx: Attempt receiving more than 1 byte

### PR DESCRIPTION
**Pull request by jori-nordic**
To not bother the CPU too much, I'd like to try receiving the number of
bytes requested through uart_fifo_read() in one DMA transaction.

The rx buffer size is set to the configured tx buffer size in order to
not introduce yet another Kconfig option that will have to be duplicated
for each peripheral instance.

Signed-off-by: Jonathan Rico <jonathan.rico@nordicsemi.no>

### CODE REVIEW

**Comment by jori-nordic** [comment]
still no reviews here, so I'll just close it. Maybe reopen later if needed.

**Comment by bamkrs** [comment]
I'm very interested in this feature. Would this pick up the whole internal hardware FIFO as well? E.g. when a high priority ISR is blocking the CPU for so long, that two bytes are received in the hardware FIFO.

**Comment by jori-nordic** [comment]
Yeah I think this results in the HW buffers being actually used. You will then be able to control the amount of bytes in the RX DMA buffer with a kconfig.

**Comment by bamkrs** [comment]
This would be super useful! +1 for having this.

**Comment by kmeinhar** [comment]
+1 from me too

File: drivers/serial/uart_nrfx_uarte.c [hunk 1]
```diff
@@ -1611,10 +1618,33 @@ static int uarte_nrfx_fifo_read(const struct device *dev,
...
+		       num_rx);
+		data->int_driven->rx_index = 0;
+
+		if (size > num_rx) {
+			/* Attempt reading the remaining bytes in a single transfer. */
+			nrf_uarte_rx_buffer_set(uarte,
+						data->int_driven->rx_buffer,
+						MIN(data->int_driven->rx_buff_size,
+						    size - num_rx));
+			nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
```
Comment (anangl) [inline drivers/serial/uart_nrfx_uarte.c:1636]: The problem with this is that a new RX interrupt will not be generated until this buffer is completely filled up and `uart_irq_rx_ready()` will be returning 0 until that happens. And this is not the way the interrupt driven API is supposed to work. https://github.com/zephyrproject-rtos/zephyr/blob/8b4407ab7cfb8319c3bf57cb01cbcb8cf9b7eb46/include/zephyr/drivers/uart.h#L1008-L1011
When a user calls for example `uart_fifo_read(dev, buf, 8)`, it does not necessarily mean that 8 characters is actually needed. It only means that at most 8 characters can be written to the passed buffer. See also:
https://github.com/zephyrproject-rtos/zephyr/blob/8b4407ab7cfb8319c3bf57cb01cbcb8cf9b7eb46/include/zephyr/drivers/uart.h#L772-L775

If you want to make use of DMA, use the asynchronous API instead (that's actually the reason why this API was introduced).

Comment (jori-nordic) [reply drivers/serial/uart_nrfx_uarte.c:1636]: thanks for the explanation. So what you're saying is that the zephyr interrupt-driven uart API will only ever allow 1 byte to be read at a time, and no DMA usage, correct? As in, this is not a limitation of our implementation or HW, just a limitation of the API itself.

**Pull request closed without merging**